vm/declare/src/lib.rs

```diff
@@ -0,0 +1,103 @@
+extern crate proc_macro;
+use proc_macro::TokenStream;
+use quote::{format_ident, quote};
+use syn::{parse_macro_input, FnArg, ImplItem, ItemImpl, LitStr, Pat};
+
+#[proc_macro_attribute]
+pub fn template(_attr: TokenStream, item: TokenStream) -> TokenStream {
+  let input = parse_macro_input!(item as ItemImpl);
+  let struct_name = &input.self_ty;
+
+  let mut c_functions = vec![];
+
+  for item in &input.items {
+    if let ImplItem::Fn(method) = item {
+      if method
+        .attrs
+        .iter()
+        .any(|attr| attr.path().is_ident("callable"))
+      {
+        let method_name = &method.sig.ident;
+        let c_func_name = format_ident!("athexp_{}", method_name);
+
+        let (params, args): (Vec<_>, Vec<_>) = method
+          .sig
+          .inputs
+          .iter()
+          .filter_map(|arg| {
+            if let FnArg::Typed(pat_type) = arg {
+              if let Pat::Ident(pat_ident) = &*pat_type.pat {
+                let ident = &pat_ident.ident;
+                let ty = &pat_type.ty;
+                Some((quote!(#ident: #ty), quote!(#ident)))
+              } else {
+                None
+              }
+            } else {
+              None
+            }
```

Comment:
This is a bit hard to follow because of many branches. How about simplifying:
```suggestion
            if let FnArg::Typed(pat_type) = arg {
              if let Pat::Ident(pat_ident) = &*pat_type.pat {
                let ident = &pat_ident.ident;
                let ty = &pat_type.ty;
                return Some((quote!(#ident: #ty), quote!(#ident)));
              }
            }
            None
```